File: swift/AntennaProtocol/Sources/AntennaProtocol/EIP712.swift
import Foundation

public enum EIP712 {
    // Matches the repository’s interop vectors (see python reference).
    private static let domainType = "EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)"
    private static let eventType = "MBEvent(bytes32 eventHash)"

    public struct Domain {
        public let name: String
        public let version: String
        public let chainId: UInt64
        public let verifyingContract: Data // 20 bytes

        public init(name: String, version: String, chainId: UInt64, verifyingContract: Data) throws {
            guard verifyingContract.count == 20 else { throw MBP2PError.invalidHex("verifyingContract must be 20 bytes") }
            self.name = name
            self.version = version
            self.chainId = chainId
            self.verifyingContract = verifyingContract
        }
    }

    public static func domainSeparator(_ domain: Domain) -> Data {
        let typeHash = Keccak256.hash(Data(domainType.utf8))
        let nameHash = Keccak256.hash(Data(domain.name.utf8))
        let versionHash = Keccak256.hash(Data(domain.version.utf8))

        var enc = Data()
        enc.append(typeHash)
        enc.append(pad32(nameHash))
        enc.append(pad32(versionHash))
        enc.append(abiEncodeUInt256(domain.chainId))
        enc.append(abiEncodeAddress(domain.verifyingContract))
        return Keccak256.hash(enc)
    }

    public static func structHashMBEvent(eventHashBytes32: Data) throws -> Data {
        guard eventHashBytes32.count == 32 else { throw MBP2PError.invalidHex("eventHash must be 32 bytes") }
        let typeHash = Keccak256.hash(Data(eventType.utf8))
        var enc = Data()
        enc.append(typeHash)
        enc.append(eventHashBytes32)
        return Keccak256.hash(enc)
    }

    public static func digest(domainSeparator: Data, structHash: Data) throws -> Data {
        guard domainSeparator.count == 32, structHash.count == 32 else {
            throw MBP2PError.invalidJSON("domainSeparator/structHash must be 32 bytes")
        }
        var data = Data([0x19, 0x01])
        data.append(domainSeparator)
        data.append(structHash)
        return Keccak256.hash(data)
    }

    // MARK: - ABI helpers (EIP-712 encoding profile)

    private static func pad32(_ data: Data) -> Data {
        if data.count == 32 { return data }
        if data.count > 32 { return data.suffix(32) }
        var out = Data(repeating: 0, count: 32 - data.count)
        out.append(data)
        return out
    }

    private static func abiEncodeUInt256(_ x: UInt64) -> Data {
        var out = Data(repeating: 0, count: 24)
        var v = x.bigEndian
        withUnsafeBytes(of: &v) { out.append(contentsOf: $0) }
        return out // 32 bytes total
    }

    private static func abiEncodeAddress(_ addr20: Data) -> Data {
        var out = Data(repeating: 0, count: 12)
        out.append(addr20)
        return out
    }
}

File: rust/antenna-protocol/src/eip712.rs
use crate::error::MBP2PError;
use crate::eip191::keccak256;

/// Matches the repository interop vectors:
/// - domain type: EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)
/// - struct type: MBEvent(bytes32 eventHash)
pub const DOMAIN_TYPE: &str = "EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)";
pub const EVENT_TYPE: &str = "MBEvent(bytes32 eventHash)";

#[derive(Debug, Clone)]
pub struct Domain {
    pub name: String,
    pub version: String,
    pub chain_id: u64,
    pub verifying_contract: [u8; 20],
}

fn abi_encode_uint256(x: u64) -> [u8; 32] {
    let mut out = [0u8; 32];
    out[24..32].copy_from_slice(&x.to_be_bytes());
    out
}

fn abi_encode_address(addr20: &[u8; 20]) -> [u8; 32] {
    let mut out = [0u8; 32];
    out[12..32].copy_from_slice(addr20);
    out
}

fn abi_encode_bytes32(b32: &[u8; 32]) -> [u8; 32] { *b32 }

fn abi_encode_string_hash(s: &str) -> [u8; 32] { keccak256(s.as_bytes()) }

pub fn domain_separator(domain: &Domain) -> [u8; 32] {
    let type_hash = keccak256(DOMAIN_TYPE.as_bytes());
    let name_hash = abi_encode_string_hash(&domain.name);
    let version_hash = abi_encode_string_hash(&domain.version);

    let mut enc = Vec::with_capacity(32 * 5);
    enc.extend_from_slice(&type_hash);
    enc.extend_from_slice(&abi_encode_bytes32(&name_hash));
    enc.extend_from_slice(&abi_encode_bytes32(&version_hash));
    enc.extend_from_slice(&abi_encode_uint256(domain.chain_id));
    enc.extend_from_slice(&abi_encode_address(&domain.verifying_contract));

    keccak256(&enc)
}

pub fn struct_hash_mbevent(event_hash: &[u8; 32]) -> [u8; 32] {
    let type_hash = keccak256(EVENT_TYPE.as_bytes());

    let mut enc = Vec::with_capacity(64);
    enc.extend_from_slice(&type_hash);
    enc.extend_from_slice(event_hash);
    keccak256(&enc)
}

pub fn digest(domain_sep: &[u8; 32], struct_hash: &[u8; 32]) -> [u8; 32] {
    let mut data = Vec::with_capacity(2 + 32 + 32);
    data.push(0x19);
    data.push(0x01);
    data.extend_from_slice(domain_sep);
    data.extend_from_slice(struct_hash);
    keccak256(&data)
}

pub fn parse_address(hex_addr: &str) -> Result<[u8; 20], MBP2PError> {
    let clean = hex_addr.trim_start_matches("0x");
    let bytes = hex::decode(clean).map_err(|e| MBP2PError::Hex(e.to_string()))?;
    if bytes.len() != 20 {
        return Err(MBP2PError::Hex("address must be 20 bytes".into()));
    }
    let mut out = [0u8; 20];
    out.copy_from_slice(&bytes);
    Ok(out)
}
